### PR TITLE
Improve support for special tokens

### LINKED ISSUE #1501: [Feature request] How to support special tokens in tokenizer of llama_cpp?
I want to expand my vocab from 32000 to 32002 by adding two special tokens: '\<start\>' and '\<end\>'. However, after I hacked the BPE model (appended two user-defined tokens), modified the size of relative weights of the model, and converted it to a .ggml file, I found that the special tokens were still split into several token IDs. Adjusting the scores of the special tokens did not work. I do not want to insert the special tokens between 0 to 32000 to keep the original order of tokens, so I have not tested inserting the special words at different positions. 

In the original SentencePiece tokenizer, it will also not work when the type of special token is not user-defined [https://github.com/google/sentencepiece/blob/master/src/sentencepiece_model.proto](sentencepiece_model.proto). So I want to ask whether llama_cpp considers supporting user-defined tokens? (Before this, an option is to use sentencepiece of python and llama-cpp-python to inference)

Thanks for your reply.

(By the way, the Llama tokenizer (BPE) was trained by adding an add_dummy_prefix option, so do not directly use the add_special_tokens function of the tokenizer of hugging face transformers in your training. Because it will first cut the entire sentence according to the special token, and then throw the remaining parts to the bpe model, so spaces will be added at the beginning of each part by the bpe model, which will be inconsistent with the token during llama_cpp inference.)

### LINKED ISSUE #1812: Vicuna 1.1 / special_tokens_map.json support
As far as I can tell, Vicuna 1.1 uses `</s>` as the separator for dialogue responses. It's tokenized as the EOS token (2) when I tried it in a Python script using Transformers, but it's tokenized as a normal string when using llama.cpp. I got something like this instead:

```
main: prompt: ' </s>'
main: number of tokens in prompt = 4
     1 -> ''
  1533 -> ' </'
 29879 -> 's'
 29958 -> '>'
```

I found these docs when looking for a reference of how the prompt should look like for Vicuna 1.1, to check if `</s>` should appear in the prompt:
https://github.com/lm-sys/FastChat/blob/7ae721fa3c881e1e24cf181305d127a316acd463/docs/vicuna_weights_version.md#example-prompt-weight-v11

```
A chat between a user and an assistant.

USER: Hello!
ASSISTANT: Hello!</s>
USER: How are you?
ASSISTANT: I am good.</s>
```

The docs at the end mention a special_tokens_map.json file that has something like this, but it doesn't seem to be used by convert.py:
``` json
  "eos_token": {
    "content": "</s>",
    "lstrip": false,
    "normalized": true,
    "rstrip": false,
    "single_word": false
  }
```

### PR DESCRIPTION
**Pull request by Igoorx**
Fixes #1812
Fixes #1501

Hello! This is my first attempt to contribute to this project, so I apologize in advance for any mistakes.
This PR should add a basic support for special tokens and improve the support for added tokens. All special tokens come from the file `added_tokens.json` and the file `special_tokens_map.json` or the file `tokenizer_config.json` ~~(I have no idea if it's safe to rely on only one so I added tokenizer_config.json as a fallback).~~
**EDIT**: The loading of the jsons now seems to follow the [huggingface implementation](https://github.com/huggingface/transformers/blob/c5454eba9eac00a3e7d0a46a3d25aacd43187f1e/src/transformers/tokenization_utils_base.py#LL1839C1-L1839C1) a bit better.

The most important points of this PR are:
- ~~The GGML format was changed due to the requirement of a way to know which tokens are the special tokens.~~
**EDIT**: This isn't necessary anymore.
- ~~The tokenizer now uses a trie algorithm to efficiently split the prompt based on the special tokens, this was necessary because the BPE tokenizer isn't able to tokenize the special tokens by itself.
Please note that this algorithm was ported from the [huggingface/transformers repository](https://github.com/huggingface/transformers/blob/ee88ae59940fd4b2c8fc119373143d7a1175c651/src/transformers/tokenization_utils.py#L52), so I wonder if this could cause license issues?~~
**EDIT**: The algorithm now is just a linear search.

Using this PR, this is the output of `--verbose-prompt`:
```
main: prompt: ' One Two</s>Three</s> Four '
main: number of tokens in prompt = 8
     1 -> '<s>'
  3118 -> ' One'
  7803 -> ' Two'
     2 -> '</s>'
 28575 -> 'Three'
     2 -> '</s>'
 12458 -> ' Four'
 29871 -> ' '
 ```
when the model is converted to ggml with this `special_tokens_map.json`:
```json
{
    "bos_token": {
      "content": "<s>",
      "lstrip": false,
      "normalized": true,
      "rstrip": false,
      "single_word": false
    },
    "eos_token": {
      "content": "</s>",
      "lstrip": false,
      "normalized": true,
      "rstrip": false,
      "single_word": false
    },
    "pad_token": "<unk>",
    "unk_token": {
      "content": "<unk>",
      "lstrip": false,
      "normalized": true,
      "rstrip": false,
      "single_word": false
    }
  }
  ```

### CODE REVIEW

**Comment by KerfuffleV2** [comment]
A breaking change to the GGML format might be a tough sell (but don't take my personal opinion as speaking for the project in any way). You might consider allowing a commandline option and/or API addition to allow just reading the special tokens from a separate file or list.

> so I wonder if this could cause license issues?

llama.cpp is under MIT and transformers seems to be Apache 2.0. I'm not qualified to say what the issue is or how to fix it, but the projects do have different licenses. I don't know if there's any kind of policy for dealing with that situation already in place. Perhaps someone else can give you a better answer, but until that part is resolved I'd guess you shouldn't expect your PR to get merged (again, not speaking with any kind of authority).

**Comment by Igoorx** [comment]
> A breaking change to the GGML format might be a tough sell (but don't take my personal opinion as speaking for the project in any way).
>You might consider allowing a commandline option and/or API addition to allow just reading the special tokens from a separate file or list.

On the bright side, the old format is still supported... But I agree with you and it's also something I would want to avoid, but the thing is, after looking at the code it looks like everything related to the vocab is integrated inside the ggml (e.g. the file `added_tokens.json` and the whole vocab itself), so I thought the only right choice was to include the list of special tokens too.
I don't know what the maintainers of the project would prefer though, so I'm open to making any changes.

> llama.cpp is under MIT and transformers seems to be Apache 2.0. I'm not qualified to say what the issue is or how to fix it, but the projects do have different licenses. I don't know if there's any kind of policy for dealing with that situation already in place.

Yes, this is something that concerns me a bit. I would appreciate feedback on whether I should be concerned about the license. After all, I'm not simply copy-pasting the code. Perhaps, the comment I added with the link to the original implementation would be sufficient?
Nevertheless, even if the license still should be respected to the fullest, I am aware that [Apache is compatible with MIT](https://opensource.stackexchange.com/questions/11640/can-mit-and-apache-licenses-be-used-together). Therefore, I believe there should be no issue with including that code in the repository. However, I also need feedback on this matter. This seems to be the first time this issue has arisen here, so I'm not sure whether the maintainers would want to add the license file of Hugging Face specifically for this piece of code. If they do, I'm not sure where I should add the license - perhaps just appending it to the LICENSE file?

> Perhaps someone else can give you a better answer, but until that part is resolved I'd guess you shouldn't expect your PR to get merged (again, not speaking with any kind of authority).

You're absolutely right, no doubts about that, that's why I brought that up in the PR message, this is something quite important for an open-source project so it needs to be sorted out before the PR is merged.

File: llama-util.h [hunk 1]
```diff
@@ -487,4 +489,166 @@ struct llama_ctx_buffer {
 typedef llama_buffer llama_ctx_buffer;
 #endif
 
+struct llama_trie_node {
+    llama_trie_node(): is_terminator(false) {}
+    
+    std::unordered_map<char, llama_trie_node*> children;
```
Comment (bullno1) [inline llama-util.h:495]: These are not freed when the node dies. Use unique_ptr?

File: llama-util.h [hunk 2]
```diff
@@ -487,4 +489,166 @@ struct llama_ctx_buffer {
...
+                break;
+            }
+        }
+        
+        offsets.push_back(text.size());
+        return offsets;
+    }
+
+private:
+    llama_trie_node *root_;
```
Comment (bullno1) [inline llama-util.h:651]: This is not freed when the trie is released, use unique_ptr?

**Review by bullno1 — COMMENTED** [review]
AFAIK, there is a discussion on a new format here: https://github.com/ggerganov/ggml/issues/220

You may want to chime in.

**Comment by Igoorx** [comment]
>AFAIK, there is a discussion on a new format here: https://github.com/ggerganov/ggml/issues/220

>
>You may want to chime in.

I don't believe I have anything to contribute to the discussion... but upon a closer look, it appears there is also a discussion regarding the tokenizer in the issue 🤔
I'm uncertain whether there is a possibility of this PR being merged or if the maintainers would prefer to wait for GGUF.
I suppose I will modify the approach of this PR and alter the GGML format in such a manner that the new model format remains compatible with older versions. By doing so, if we still have doubts about when GGUF will be merged, this pull request could serve as a temporary solution to the issue of special tokens at least until that time.

**Comment by grantbey** [comment]
Hey @Igoorx what's the status of this PR? I'm really interested in this work.

I fine tune using LoRA and add a few special tokens, but then these aren't tokenized correctly when running inference on llama.cpp. I'm going to try your PR and see if it helps things.

**Comment by grantbey** [comment]
Thanks. I’ve started taking a stab at the merge conflicts but I imagine
I’ll get a few things wrong and it’ll end up consuming a lot of my time 😅

I read about GGUF but I’ve got some deadlines coming up and I’d like a
working solution sooner rather than later, and it’s not clear what kind of
timeline to expect GGUF on.

**Comment by Igoorx** [comment]
@grantbey I rebased the PR to the last master commit :+1:

**Comment by grantbey** [comment]
Ok wow @Igoorx you're amazing. Thank you so much!

**Comment by goerch** [comment]
@Igoorx : @klosax just made me aware that we are working on weaknesses of the tokenizer(s) at the same time. I'd greatly appreciate any cooperation on this. If I'd go to incorporate your changes my first question would be how to test them?

**Comment by Igoorx** [comment]
@goerch
Here is a simple test: https://github.com/ggerganov/llama.cpp/pull/1931/commits/6f7dabab441566078446ef868e573cd309fe62be

**Comment by goerch** [comment]
> @goerch Here is a simple test: [6f7daba](https://github.com/ggerganov/llama.cpp/commit/6f7dabab441566078446ef868e573cd309fe62be)

Nice, thank you! Then my plan would be to try to first integrate your changes into #2315 and afterwards migrate the PR over to `gguf`. Would that be OK for you?

**Comment by Igoorx** [comment]
@goerch Yeah, that's fine. You just have to be aware about: https://github.com/ggerganov/llama.cpp/blob/6f7dabab441566078446ef868e573cd309fe62be/llama-util.h#L554-L555
The trie algorithm used in the PR is a port from the huggingface repository, as written in the comment, so maybe something needs to be done about it. I'm not sure if that comment is enough or if it would be necessary to add the hf license somewhere.

**Comment by klosax** [comment]
The gguf gpt2 tokenizer also have a Trie implementation. The tokenizer is on MIT license. Maybe it could be reused for the llama tokenizer.

**Comment by goerch** [comment]
> @goerch Yeah, that's fine. You just have to be aware about: ...

IANAL, but I'm equally concerned about compatibility with the `sentencepiece` license (Apache-2.0). They use a trie too, which might the base of the `hf` implementation? When developing #2315 I experimented with a simple clean room implementation of a trie already, I'll try to understand the differences.

@ggerganov and others: any opinions on this?

**Comment by klosax** [comment]
The author of the gpt2 tokenizer gave permission to use it and stated it is on MIT license here https://github.com/ggerganov/llama.cpp/pull/2398#issuecomment-1667009979

**Comment by Igoorx** [comment]
The important part is the `split` method
https://github.com/ggerganov/llama.cpp/blob/6f7dabab441566078446ef868e573cd309fe62be/llama-util.h#L575-L577
If using the ported version isn't an option, would be necessary to reimplement it using another trie algorithm.

>IANAL, but I'm equally concerned about compatibility with the sentencepiece license (Apache-2.0).

IANAL either, but Apache-2.0 and MIT should be compatible with each other: https://law.stackexchange.com/a/6732

**Comment by klosax** [comment]
It looks like the MIT and Apache licenses are compatible, but a copy of the Apache license and a Notice file must be included:
https://softwareengineering.stackexchange.com/questions/51987/how-to-include-an-apache-library-with-my-opensource-code#52223

**Comment by ggerganov** [comment]
I'll recommend to either implement a trie from scratch, or use a linear search algorithm - we are not tokenizing billions of tokens, so not sure what we gain from using a trie.

**Comment by goerch** [comment]
@Igoorx : I took a look into your PR, but I believe we first have to sort out open problems with #2549.  Sorry for the delay.

**Comment by alitariq4589** [comment]
Can one of the admins verify this patch?

**Comment by goerch** [comment]
@lgoorx : #2549 is done, I'm now waiting for review of https://github.com/ggerganov/llama.cpp/pull/3252 and am already downloading Vicuna-7B-v1.1.

**Comment by goerch** [comment]
Currently discussing this at https://github.com/ggerganov/llama.cpp/issues/2820.

**Comment by Igoorx** [comment]
Looks like this PR was superseded by https://github.com/ggerganov/llama.cpp/pull/3538, from what I could see it looks great. Thanks for your attention @goerch! I don't think I have anything more to contribute.

**Pull request closed without merging**